File: packages/dotnet/NationalIdentifiers.Core/Countries/ItalyIdentity.cs
using System.Globalization;
using System.Text;

namespace NationalIdentifiers.Core.Countries;

internal static class ItalyIdentity
{
    private const string Vowels = "AEIOU";
    private const string MonthLetters = "ABCDEHLMPRST";

    // Positions that hold digits in a non-omocodic fiscal code.
    private static readonly int[] NumericPositions = [6, 7, 9, 10, 12, 13, 14];

    private static readonly IReadOnlyDictionary<char, char> OmocodiaDecode =
        new Dictionary<char, char>
        {
            ['L'] = '0', ['M'] = '1', ['N'] = '2', ['P'] = '3', ['Q'] = '4',
            ['R'] = '5', ['S'] = '6', ['T'] = '7', ['U'] = '8', ['V'] = '9',
        };

    internal static (IReadOnlyList<string> Checked, IReadOnlyList<string> Mismatched) Check(
        string normalizedTaxId,
        TaxIdIdentity identity)
    {
        var checkedFields = new List<string>();
        var mismatchedFields = new List<string>();
        var code = DecodeOmocodia(normalizedTaxId);

        int encodedDay = int.Parse(code[9..11], CultureInfo.InvariantCulture);
        int dayOfBirth = encodedDay > 40 ? encodedDay - 40 : encodedDay;

        if (!string.IsNullOrWhiteSpace(identity.LastName))
        {
            checkedFields.Add("lastName");
            if (EncodeSurname(identity.LastName) != code[..3])
                mismatchedFields.Add("lastName");
        }

        if (!string.IsNullOrWhiteSpace(identity.FirstName))
        {
            checkedFields.Add("firstName");
            if (EncodeGivenName(identity.FirstName) != code[3..6])
                mismatchedFields.Add("firstName");
        }

        if (identity.BirthDate is { } birthDate)
        {
            checkedFields.Add("birthDate");
            bool yearMatches = code[6..8] ==
                (birthDate.Year % 100).ToString("D2", CultureInfo.InvariantCulture);
            bool monthMatches = code[8] == MonthLetters[birthDate.Month - 1];
            bool dayMatches = dayOfBirth == birthDate.Day;
            if (!yearMatches || !monthMatches || !dayMatches)
                mismatchedFields.Add("birthDate");
        }

        if (identity.Gender is 'M' or 'F')
        {
            checkedFields.Add("gender");
            char encodedGender = encodedDay > 40 ? 'F' : 'M';
            if (identity.Gender != encodedGender)
                mismatchedFields.Add("gender");
        }

        if (!string.IsNullOrWhiteSpace(identity.BirthPlaceCode))
        {
            checkedFields.Add("birthPlaceCode");
            if (identity.BirthPlaceCode.Trim().ToUpperInvariant() != code[11..15])
                mismatchedFields.Add("birthPlaceCode");
        }

        return (checkedFields, mismatchedFields);
    }

    internal static string DecodeOmocodia(string fiscalCode)
    {
        var characters = fiscalCode.ToCharArray();

        foreach (int position in NumericPositions)
        {
            if (OmocodiaDecode.TryGetValue(characters[position], out char decoded))
                characters[position] = decoded;
        }

        return new string(characters);
    }

    internal static string EncodeSurname(string surname)
    {
        var (consonants, vowels) = SplitLetters(NormalizeNamePart(surname));
        return PadCode(consonants + vowels);
    }

    internal static string EncodeGivenName(string givenName)
    {
        var (consonants, vowels) = SplitLetters(NormalizeNamePart(givenName));

        if (consonants.Length >= 4)
            return string.Concat(consonants[0], consonants[2], consonants[3]);

        return PadCode(consonants + vowels);
    }

    private static string PadCode(string letters) =>
        (letters.Length >= 3 ? letters[..3] : letters).PadRight(3, 'X');

    private static string NormalizeNamePart(string value)
    {
        var builder = new StringBuilder(value.Length);

        foreach (char c in value.Normalize(NormalizationForm.FormD))
        {
            if (CharUnicodeInfo.GetUnicodeCategory(c) == UnicodeCategory.NonSpacingMark)
                continue;

            char upper = char.ToUpperInvariant(c);
            if (upper is >= 'A' and <= 'Z')
                builder.Append(upper);
        }

        return builder.ToString();
    }

    private static (string Consonants, string Vowels) SplitLetters(string letters)
    {
        var consonants = new StringBuilder();
        var vowels = new StringBuilder();

        foreach (char letter in letters)
        {
            if (Vowels.Contains(letter))
                vowels.Append(letter);
            else
                consonants.Append(letter);
        }

        return (consonants.ToString(), vowels.ToString());
    }
}

File: projects/tax-id/src/lib/countries/italy-identity.ts
import type { TaxIdIdentity, TaxIdIdentityField } from '../identity-consistency';

const VOWELS = 'AEIOU';
const MONTH_LETTERS = 'ABCDEHLMPRST';

// Omocodia substitution table (digit -> letter), inverted for decoding.
const OMOCODIA_DECODE: Readonly<Record<string, string>> = {
  L: '0', M: '1', N: '2', P: '3', Q: '4',
  R: '5', S: '6', T: '7', U: '8', V: '9',
};

// Positions that hold digits in a non-omocodic fiscal code.
const NUMERIC_POSITIONS = [6, 7, 9, 10, 12, 13, 14] as const;

function normalizeNamePart(value: string): string {
  return value
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .toUpperCase()
    .replace(/[^A-Z]/g, '');
}

function splitLetters(value: string): { consonants: string; vowels: string } {
  let consonants = '';
  let vowels = '';

  for (const letter of value) {
    if (VOWELS.includes(letter)) {
      vowels += letter;
    } else {
      consonants += letter;
    }
  }

  return { consonants, vowels };
}

export function encodeItalianSurname(surname: string): string {
  const { consonants, vowels } = splitLetters(normalizeNamePart(surname));
  return (consonants + vowels).slice(0, 3).padEnd(3, 'X');
}

export function encodeItalianGivenName(givenName: string): string {
  const { consonants, vowels } = splitLetters(normalizeNamePart(givenName));

  if (consonants.length >= 4) {
    return consonants[0] + consonants[2] + consonants[3];
  }

  return (consonants + vowels).slice(0, 3).padEnd(3, 'X');
}

/** Restores the digits replaced by omocodia substitutions. */
export function decodeItalianOmocodia(fiscalCode: string): string {
  const characters = fiscalCode.split('');

  for (const position of NUMERIC_POSITIONS) {
    const decoded = OMOCODIA_DECODE[characters[position]];
    if (decoded !== undefined) {
      characters[position] = decoded;
    }
  }

  return characters.join('');
}

export function checkItalianIdentity(
  normalizedTaxId: string,
  identity: TaxIdIdentity,
): { checked: TaxIdIdentityField[]; mismatched: TaxIdIdentityField[] } {
  // Biographical data is encoded only in the 16-character personal fiscal
  // code; an 11-digit partita IVA carries none, so nothing is checkable.
  if (normalizedTaxId.length !== 16) {
    return { checked: [], mismatched: [] };
  }

  const checked: TaxIdIdentityField[] = [];
  const mismatched: TaxIdIdentityField[] = [];
  const code = decodeItalianOmocodia(normalizedTaxId);

  const encodedDay = Number(code.slice(9, 11));
  const dayOfBirth = encodedDay > 40 ? encodedDay - 40 : encodedDay;

  if (identity.lastName) {
    checked.push('lastName');
    if (encodeItalianSurname(identity.lastName) !== code.slice(0, 3)) {
      mismatched.push('lastName');
    }
  }

  if (identity.firstName) {
    checked.push('firstName');
    if (encodeItalianGivenName(identity.firstName) !== code.slice(3, 6)) {
      mismatched.push('firstName');
    }
  }

  const birthDate = parseIsoDate(identity.birthDate);
  if (birthDate) {
    checked.push('birthDate');
    const yearMatches = code.slice(6, 8) === String(birthDate.year % 100).padStart(2, '0');
    const monthMatches = code[8] === MONTH_LETTERS[birthDate.month - 1];
    const dayMatches = dayOfBirth === birthDate.day;
    if (!yearMatches || !monthMatches || !dayMatches) {
      mismatched.push('birthDate');
    }
  }

  if (identity.gender === 'M' || identity.gender === 'F') {
    checked.push('gender');
    const encodedGender = encodedDay > 40 ? 'F' : 'M';
    if (identity.gender !== encodedGender) {
      mismatched.push('gender');
    }
  }

  if (identity.birthPlaceCode) {
    checked.push('birthPlaceCode');
    if (identity.birthPlaceCode.trim().toUpperCase() !== code.slice(11, 15)) {
      mismatched.push('birthPlaceCode');
    }
  }

  return { checked, mismatched };
}

function parseIsoDate(
  value: string | undefined,
): { year: number; month: number; day: number } | null {
  if (!value || !/^\d{4}-\d{2}-\d{2}$/.test(value)) {
    return null;
  }

  const year = Number(value.slice(0, 4));
  const month = Number(value.slice(5, 7));
  const day = Number(value.slice(8, 10));
  const candidate = new Date(Date.UTC(year, month - 1, day));

  const isRealDate =
    candidate.getUTCFullYear() === year &&
    candidate.getUTCMonth() === month - 1 &&
    candidate.getUTCDate() === day;

  return isRealDate ? { year, month, day } : null;
}
